### PR TITLE
feat: Add Sentry Babel Transformer

### PR DESCRIPTION
**Pull request by krystofwoldrich**
## :loudspeaker: Type of change

- [x] New feature

## :scroll: Description

This PR adds `@sentry/babel-plugin-component-annotate` to `@sentry/react-native/metro`.

- Similar to the rest of our JS Bundlers Plugins https://github.com/getsentry/sentry-javascript-bundler-plugins/pull/469

Example use RN 0.72 and newer:

```js
// For Expo
const { getSentryExpoConfig } = require("@sentry/react-native/metro");
const config = getSentryExpoConfig(__dirname, { annotateReactComponents: true });

// For RN
const m = mergeConfig(getDefaultConfig(__dirname), config);

module.exports = withSentryConfig(m, {
  annotateReactComponents: true, // default false
});
```

For RN 0.65 to 0.71 (these version did not have default config in metro.config.js but applied it later, that changed in 0.72):
```
const config  = {
  transformer: {
	babelTransformerPath: require.resolve('metro-babel-transformer'),
	// this must be adjust to appropriate babel transformer for given RN version
  },
};
```

## :bulb: Motivation and Context

- closes: https://github.com/getsentry/sentry-react-native/issues/3913

## :green_heart: How did you test it?
sample apps, e2e tests, unit tests

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled
- [x] All tests passing
- [x] No breaking changes

## :crystal_ball: Next steps
- [ ] Add docs

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,18 @@
 - Improve touch event component info if annotated with [`@sentry/babel-plugin-component-annotate`](https://www.npmjs.com/package/@sentry/babel-plugin-component-annotate) ([#3899](https://github.com/getsentry/sentry-react-native/pull/3899))
```
Comment (vaind) [inline CHANGELOG.md:7]: maybe remove this from the changelog?

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,18 @@
...
 - Add replay breadcrumbs for touch & navigation events ([#3846](https://github.com/getsentry/sentry-react-native/pull/3846))
 - Add network data to Session Replays ([#3912](https://github.com/getsentry/sentry-react-native/pull/3912))
+- Add `annotateReactComponents` option to `@sentry/react-native/metro` ([#3916](https://github.com/getsentry/sentry-react-native/pull/3916))
```
Comment (vaind) [inline CHANGELOG.md:10]: How about just `annotateComponents`?

**Review by vaind — COMMENTED** [review]
This goes well beyond my understanding of the ecosystem but the I approve of the idea, because it's exactly what I wanted to do at first, I just didn't know how.

File: CHANGELOG.md (hunk 1 above)
Comment (krystofwoldrich) [reply CHANGELOG.md:7]: Thanks good point, I'll update the changelog.

File: src/js/tools/sentryBabelTransformer.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
+import componentAnnotatePlugin from '@sentry/babel-plugin-component-annotate';
+
+import { enableLogger } from './enableLogger';
+import { loadDefaultBabelTransformer } from './sentryBabelTransformerUtils';
+import type { BabelTransformer, BabelTransformerArgs } from './vendor/metro/metroBabelTransformer';
+
+enableLogger();
```
Comment (lforst) [inline src/js/tools/sentryBabelTransformer.ts:7]: Do we for sure want this call here?

File: src/js/tools/metroconfig.ts [hunk 4]
```diff
@@ -64,6 +96,38 @@ function loadExpoMetroConfigModule(): {
   }
 }
 
+/**
+ * Adds Sentry Babel transformer to the Metro config.
+ */
+export function withSentryBabelTransformer(config: MetroConfig): MetroConfig {
+  const defaultBabelTransformerPath = config.transformer && config.transformer.babelTransformerPath;
+  logger.debug('Default Babel transformer path from `config.transformer`:', defaultBabelTransformerPath);
```
Comment (lforst) [inline src/js/tools/metroconfig.ts:104]: The Sentry SDK logger is mostly intended to be used when the SDK is initialized. For build tooling we usually don't rely on it. I would do a good ol' console.log instead.

File: src/js/tools/metroconfig.ts [hunk 5]
```diff
@@ -64,6 +96,38 @@ function loadExpoMetroConfigModule(): {
...
+    process.on('exit', () => {
+      cleanDefaultBabelTransformerPath();
+    });
+  }
+
+  return {
+    ...config,
+    transformer: {
+      ...config.transformer,
+      babelTransformerPath: require.resolve('./sentryBabelTransformer'),
```
Comment (lforst) [inline src/js/tools/metroconfig.ts:126]: I don't know about the state of ESM/CJS in react native but `require.resolve` will not work in ESM.

File: src/js/tools/sentryBabelTransformerUtils.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+/**
+ * Reads default Babel transformer path from the project root.
+ */
+export function readDefaultBabelTransformerPath(): string | undefined {
+  try {
+    if (fs.existsSync(getDefaultBabelTransformerPath())) {
+      return fs.readFileSync(getDefaultBabelTransformerPath()).toString();
+    }
```
Comment (lforst) [inline src/js/tools/sentryBabelTransformerUtils.ts:29]: I think this pattern of fs.exists then fs.readFile should be avoided. It technically is a race condition. readFile throws anyhow if the file doesn't exist.

**Review by lforst — APPROVED** [review]
Take with a grain of salt because I am no expert but nothing in this PR looks completely off to me.

File: src/js/tools/sentryBabelTransformer.ts (hunk 3 above)
Comment (krystofwoldrich) [reply src/js/tools/sentryBabelTransformer.ts:7]: It only enables it when `env.SENTRY_LOG_LEVEL === 'debug'`. The babelTransformer is created when the file loads, so I call it here not to miss any logs (if enabled).

File: src/js/tools/metroconfig.ts (hunk 5 above)
Comment (krystofwoldrich) [reply src/js/tools/metroconfig.ts:126]: The `@sentry/react-native/metro` is CJS, same as the `metro` package.

**Review by lucas-zimerman — APPROVED** [review]
Looks good to me from my part, LGTM!

**Comment by 0Calories** [comment]
Thanks for expanding support to react native! This looks great

**Pull request merged**